File: triggers/src/main/frontend/flash-action.ts
// Client-side factory for FlashAction (custom action — UC5).
// Pairs with com.example.uc5.FlashAction on the server. Registers a
// "demo:flash" action against window.Vaadin.Flow.triggers, which is the
// registry installed by flow-client/Triggers.ts.

interface TriggersRegistry {
  registerAction(
    typeId: string,
    factory: (
      config: Record<string, unknown>,
      extras: HTMLElement[]
    ) => { run(): void }
  ): void;
}

const triggers = (window as unknown as {
  Vaadin?: { Flow?: { triggers?: TriggersRegistry } };
}).Vaadin?.Flow?.triggers;

if (!triggers) {
  console.debug(
    'window.Vaadin.Flow.triggers not available — flash-action.ts loaded too early'
  );
} else {
  triggers.registerAction('demo:flash', (config, extras) => {
    const elementIndex = Number(config.element ?? 0);
    return {
      run() {
        // elementIndex 0 means "host"; this action only supports an extra
        // target. extras[i] is the (i+1)-th referenced element.
        const target =
          elementIndex === 0 ? null : extras[elementIndex - 1] ?? null;
        if (!target) {
          return;
        }
        const originalBg = target.style.backgroundColor;
        const originalTransition = target.style.transition;
        target.style.transition = 'background-color 200ms';
        target.style.backgroundColor = 'var(--aura-yellow, gold)';
        window.setTimeout(() => {
          target.style.backgroundColor = originalBg;
          window.setTimeout(() => {
            target.style.transition = originalTransition;
          }, 220);
        }, 220);
      }
    };
  });
}

export {};

File: triggers/src/main/java/com/example/uc5/FlashAction.java
package com.example.uc5;

import java.util.Objects;

import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.trigger.internal.Action;
import com.vaadin.flow.component.trigger.internal.Trigger;
import com.vaadin.flow.dom.Element;
import com.vaadin.flow.dom.JsFunction;

/**
 * Custom action that briefly flashes the target element's background.
 * <p>
 * Demonstrates the extension SPI: subclass {@link Action} and emit the JS
 * that runs when the trigger fires by returning a {@link JsFunction} from
 * {@link #toJs(Trigger)}. The target element is captured as {@code $0} so it
 * arrives on the client as a DOM reference — no {@code @JsModule} or
 * client-side registry needed.
 */
public class FlashAction extends Action {

    private static final String BODY = """
            const t=$0;
            const o=t.style.backgroundColor;
            t.style.transition='background-color 200ms';
            t.style.backgroundColor='var(--aura-yellow, gold)';
            window.setTimeout(()=>{t.style.backgroundColor=o;},220);""";

    private final Element target;

    public FlashAction(Component target) {
        this(Objects.requireNonNull(target).getElement());
    }

    public FlashAction(Element target) {
        this.target = Objects.requireNonNull(target);
    }

    public Element getTarget() {
        return target;
    }

    @Override
    protected JsFunction toJs(Trigger trigger) {
        return JsFunction.of(BODY, target);
    }
}
